src/callbacks/stepsize.jl

```diff
@@ -0,0 +1,76 @@
+
+"""
+    StepsizeCallback(; cfl=1.0)
+
+Set the time step size according to a CFL condition with CFL number `cfl`
+if the time integration method isn't adaptive itself.
+"""
+mutable struct StepsizeCallback{RealT}
+  cfl_number::RealT
+end
+
+
+function Base.show(io::IO, cb::DiscreteCallback{Condition,Affect!}) where {Condition, Affect!<:StepsizeCallback}
+  stepsize_callback = cb.affect!
+  @unpack cfl_number = stepsize_callback
+  print(io, "StepsizeCallback(cfl_number=", cfl_number, ")")
+end
+# TODO: Taal bikeshedding, implement a method with more information and the signature
+# function Base.show(io::IO, ::MIME"text/plain", cb::DiscreteCallback{Condition,Affect!}) where {Condition, Affect!<:StepsizeCallback}
+# end
+
+
+function StepsizeCallback(; cfl::Real=1.0)
+  # when is the callback activated
+  condition = (u, t, integrator) -> true
+
+  stepsize_callback = StepsizeCallback(cfl)
+
+  DiscreteCallback(condition, stepsize_callback,
+                   save_positions=(false,false),
+                   initialize=initialize!)
+end
+
+
+function initialize!(cb::DiscreteCallback{Condition,Affect!}, u, t, integrator) where {Condition, Affect!<:StepsizeCallback}
+  cb.affect!(integrator)
+end
+
+
+@inline function (stepsize_callback::StepsizeCallback)(integrator)
+  # TODO: Taal decide, shall we set the time step even if the integrator is adaptive?
+  if !integrator.opts.adaptive
+    t = integrator.t
+    u_ode = integrator.u
+    semi = integrator.p
+    mesh, equations, solver, cache = mesh_equations_solver_cache(semi)
+    @unpack cfl_number = stepsize_callback
+    u = wrap_array(u_ode, mesh, equations, solver, cache)
```

Comment:
Correct. `u_ode` is still reachable as `integrator.u` and can't be GCed. We only need to use `GC.@preserve` if we use something like `wrap_array(copy(u_ode), mesh, equations, solver, cache)`, where the `copy` can be GCed since it's not used directly afterwards but only indirectly via the wrapped version.